### PR TITLE
STORM-1233: Port AuthUtilsTest to java

### PR DESCRIPTION
**Pull request by abellina**
I did a little bit of refactoring in AuthUtils:
- Using Hex.encodeHexString in makeDigestPayload instead of StringBuilder and String.format
- Factored out the reflective class instantiations as a generic functions.
- Added a null check in updateSubject

I am also using JUnit TemporaryFolder to create temp folders which should be deleted after the test runs.

### CODE REVIEW

File: storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java [hunk 1]
```diff
@@ -81,35 +80,88 @@ public static Configuration GetConfiguration(Map storm_conf) {
...
+     * @return Instance of type T
+     */
+    private static <T> T makeClass(String klassName) {
+        try {
+            Class klass = Class.forName(klassName);
+            return (T)klass.newInstance();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
```
Comment (hustfxj) [inline storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java]: I think it's better to replace this method by  Utils.newInstance

File: storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        AppConfigurationEntry entry = Mockito.mock(AppConfigurationEntry.class);
+        
+        Mockito.<Map<String, ?>>when(entry.getOptions()).thenReturn(optionMap);
+        String section = "bogus-section";
+        Configuration mockConfig = Mockito.mock(Configuration.class);
+        Mockito.when(mockConfig.getAppConfigurationEntry(section))
+               .thenReturn(new AppConfigurationEntry[] {entry});
+        Assert.assertEquals(
+                AuthUtils.get(mockConfig, section, "existent-key"), "foo");
+    }
```
Comment (hustfxj) [inline storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java]: The test logic is different with clojure's test-returns-first-value-for-valid-key

Comment (abellina) [inline storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java]: Thanks @hustfxj, fixed that.

File: storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java [hunk 3]
```diff
@@ -240,8 +290,8 @@ public static Subject populateSubject(Subject subject, Collection<IAutoCredentia
      * @param credentials the credentials to pull from
      */
     public static void updateSubject(Subject subject, Collection<IAutoCredentials> autos, Map<String,String> credentials) {
-        if (subject == null) {
-            throw new RuntimeException("The subject cannot be null when updating a subject with credentials");
+        if (subject == null || autos == null) {
+            throw new RuntimeException("Both subject and auto credentials cannot be null when updating a subject with credentials");
```
Comment (abhishekagarwal87) [inline storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java]: Minor. "The subject or auto credentials.."

**Comment by abellina** [comment]
Test failures appear to be unrelated to my change.

File: storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+import org.apache.storm.Config;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class AuthUtilsTest {
+
+    @Test(expected=IOException.class)
```
Comment (abhishekagarwal87) [inline storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java]: nit. space around =

File: storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        String stringFormatMethod = builder.toString();
+
+        Assert.assertEquals(
+            AuthUtils.makeDigestPayload(mockConfig, "user-pass-section"),
+            sha);
+
+        Assert.assertEquals(sha, stringFormatMethod); 
+    }
+
+    @Test(expected=RuntimeException.class)
```
Comment (redsanket) [inline storm-core/test/jvm/org/apache/storm/security/auth/AuthUtilsTest.java]: same here

**Comment by abellina** [comment]
@abhishekagarwal87 @redsanket any other changes that need made?

**Comment by HeartSaVioR** [comment]
Sorry for being late. I found several style issues, but I believe we have to decide code style guide and fix all the codes at once. If formatter can format automatically and nicely it should be great.
Anyway other things look good. +1 @abellina Thanks for the great work.

**Comment by HeartSaVioR** [comment]
Sorry I didn't see related test failing. Pulling PR to local and running build manually.

**Comment by HeartSaVioR** [comment]
Test runs fine with local (pulled recent master branch). Merging...

**Comment by abellina** [comment]
@HeartSaVioR Thanks for merging. What style issues did you find? It would be good to know to keep in mind for new pull requests. Also, what is the progress on a code style guide? I see: https://github.com/apache/storm/pull/478, but I don't think we have taken steps after that.

File: storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java [hunk 6]
```diff
@@ -81,48 +81,86 @@ public static Configuration GetConfiguration(Map storm_conf) {
...
+    public static String get(Configuration configuration, String section, String key) throws IOException {
+        AppConfigurationEntry[] configurationEntries = AuthUtils.getEntries(configuration, section);
+
+        if (configurationEntries == null){
+            return null;
+        }
+
+        for (AppConfigurationEntry entry: configurationEntries) {
+            Object val = entry.getOptions().get(key);
+            if (val != null)
```
Comment (HeartSaVioR) [inline storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java]: Trivial: We're using bracket even though it is single line.

File: storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java [hunk 7]
```diff
@@ -81,48 +81,86 @@ public static Configuration GetConfiguration(Map storm_conf) {
...
+     * Pull a the value given section and key from Configuration
+     * @param configuration The config to pull the key/value pairs out of.
+     * @param section The app configuration entry name to get stuff from.
+     * @param key The key to look up inside of the section
+     * @return Return a the String value of the configuration value
+     */
+    public static String get(Configuration configuration, String section, String key) throws IOException {
+        AppConfigurationEntry[] configurationEntries = AuthUtils.getEntries(configuration, section);
+
+        if (configurationEntries == null){
```
Comment (HeartSaVioR) [inline storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java]: Trivial: need space between ) and {

File: storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java [hunk 8]
```diff
@@ -81,48 +81,86 @@ public static Configuration GetConfiguration(Map storm_conf) {
...
      * @return the plugin
      */
     public static IPrincipalToLocal GetPrincipalToLocalPlugin(Map storm_conf) {
         IPrincipalToLocal ptol;
         try {
-          String ptol_klassName = (String) storm_conf.get(Config.STORM_PRINCIPAL_TO_LOCAL_PLUGIN);
-          Class klass = Class.forName(ptol_klassName);
-          ptol = (IPrincipalToLocal)klass.newInstance();
-          ptol.prepare(storm_conf);
+            String ptol_klassName = (String) storm_conf.get(Config.STORM_PRINCIPAL_TO_LOCAL_PLUGIN);
```
Comment (HeartSaVioR) [inline storm-core/src/jvm/org/apache/storm/security/auth/AuthUtils.java]: I think we need to convert variable name to use camel case as well. But it should be corrected from whole code after porting is complete.

**Comment by HeartSaVioR** [comment]
@abellina There're minor or spot you missed. We haven't set up a style guide yet so better to take step ASAP.

**Pull request merged**